CRM/Sepa/Logic/Queue/Close.php

```diff
@@ -56,23 +63,30 @@ public static function launchCloseRunner($txgroup_ids, $target_group_status, $ta
       // count the contributions and create an appropriate amount of items
       $contribution_count = CRM_Core_DAO::singleValueQuery("SELECT COUNT(contribution_id) FROM civicrm_sdd_contribution_txgroup WHERE txgroup_id={$txgroup['id']}");
       $contribution_count += SDD_CLOSE_RUNNER_BATCH_SIZE; // security margin
-      for ($offset=0; $offset <= $contribution_count; $offset+=SDD_CLOSE_RUNNER_BATCH_SIZE) {
+      for ($offset=0; $offset <= $contribution_count; $offset += SDD_CLOSE_RUNNER_BATCH_SIZE) {
         $queue->createItem(new CRM_Sepa_Logic_Queue_Close('update_contribution', $txgroup, $target_contribution_status, $offset));
       }
 
       // finally: render XML and mark the group
-      $queue->createItem(new CRM_Sepa_Logic_Queue_Close('create_xml', $txgroup, $target_group_status));
+      if ($is_received_runner) {
+        $queue->createItem(new CRM_Sepa_Logic_Queue_Close('create_xml', $txgroup, $target_group_status));
```

Comment:
Adding items should only be done when the queue is empty or doesn't exist. This should be checked with `$queue->existsQueue()` before adding items. Otherwise, loading the `CRM_Sepa_Page_MarkGroupReceived` page again would load the existing queue and add items for all contributions again, and also starting the runner again (which might stall either runner due to the next item being already claimed by the other).